lib/ui/painting/codec.cc

```diff
@@ -395,60 +383,44 @@ int MultiFrameCodec::repetitionCount() {
 
 sk_sp<SkImage> MultiFrameCodec::GetNextFrameImage(
     fml::WeakPtr<GrContext> resourceContext) {
-  // Populate this bitmap from the cache if it exists
-  SkBitmap bitmap = frameBitmaps_[nextFrameIndex_] != nullptr
-                        ? *frameBitmaps_[nextFrameIndex_]
-                        : SkBitmap();
-  const bool frameAlreadyCached = bitmap.getPixels();
-  if (!frameAlreadyCached) {
-    SkImageInfo info = codec_->getInfo().makeColorType(kN32_SkColorType);
-    if (info.alphaType() == kUnpremul_SkAlphaType) {
-      info = info.makeAlphaType(kPremul_SkAlphaType);
-    }
-    bitmap.allocPixels(info);
-
-    SkCodec::Options options;
-    options.fFrameIndex = nextFrameIndex_;
-    const int requiredFrameIndex = frameInfos_[nextFrameIndex_].fRequiredFrame;
-    if (requiredFrameIndex != SkCodec::kNoFrame) {
-      if (lastRequiredFrame_ == nullptr) {
-        FML_LOG(ERROR) << "Frame " << nextFrameIndex_ << " depends on frame "
-                       << requiredFrameIndex
-                       << " and no required frames are cached.";
-        return NULL;
-      } else if (lastRequiredFrameIndex_ != requiredFrameIndex) {
-        FML_DLOG(INFO) << "Required frame " << requiredFrameIndex
-                       << " is not cached. Using " << lastRequiredFrameIndex_
-                       << " instead";
-      }
-
-      if (lastRequiredFrame_->getPixels() &&
-          copy_to(&bitmap, lastRequiredFrame_->colorType(),
-                  *lastRequiredFrame_)) {
-        options.fPriorFrame = requiredFrameIndex;
-      }
-    }
-
-    if (SkCodec::kSuccess != codec_->getPixels(info, bitmap.getPixels(),
-                                               bitmap.rowBytes(), &options)) {
-      FML_LOG(ERROR) << "Could not getPixels for frame " << nextFrameIndex_;
+  SkBitmap bitmap = SkBitmap();
+  SkImageInfo info = codec_->getInfo().makeColorType(kN32_SkColorType);
+  if (info.alphaType() == kUnpremul_SkAlphaType) {
+    info = info.makeAlphaType(kPremul_SkAlphaType);
+  }
+  bitmap.allocPixels(info);
+
+  SkCodec::Options options;
+  options.fFrameIndex = nextFrameIndex_;
+  const int requiredFrameIndex = frameInfos_[nextFrameIndex_].fRequiredFrame;
```

Comment:
Would be nice to remove `frameInfos_` and the initialization loop in the constructor.

Reply:
At first I was thinking we still needed to precompute the required frames to handle cases like this (from [this blog](https://01.org/chromium/blogs/2016/chrome-graphics-internals-gif-animations-decoding-and-rendering)):

![chart](https://01.org/sites/default/files/pictures/chrome_01.jpg)

But after digging into it more, we can toss it. `fDisposalMethod` tells us whether we need to cache the frame at all or not, and previously we've learned that we only need to use the most recently cached frame and not bother tracking exact specific required frames. So I tore `requiredFrames_` out and switched to checking `fDisposalMethod` instead.